### PR TITLE
Jialfred push android sdk enhancements

### PR DESCRIPTION
**Pull request by sjimdickens**

### CODE REVIEW

**Comment by svkrish** [comment]
@sjimdickens what are these enhancements about - specs, story ??  something that speaks of the requirement that coded here.  Can you please help with that information so that I can review these in that backdrop.  Thanks.

**Comment by sjimdickens** [comment]
Hi Venkat, this is related to this [story](https://github.ibm.com/bluemix-mobile-services/bmd-project-push/issues/250)

**Comment by svkrish** [comment]
**MFPPushIntentService.java**
1. For bridging... I saw this https://developer.android.com/wear/preview/features/bridger.html.  So, should we be really supporting this ?  BTW, I just looked up to understand bridge and chanced on this :)   How are we using 'redact".
2. Can you please arrange the version checks and action in ascending order... maybe it will give a better flow for maintenance. Like check if (androidSDKVersion > 15)  ahead of if (androidSDKVersion > 19) 
**MFPInternalPushMessage.java**
_private MFPInternalPushMessage(Parcel source)_
I see you are parsing priority and visibility strings and setting the to integer literals.  Instead of using these literals why don't we use the Constants defined by android... like NotificationCompat.PRIORITY_HIGH etc.
**General**
Over I see that priority and visibility are set if they are set on the server when sending notifications.  I am unable to understand how these can be overriden by the Android App. developer.  I looked up the application code and did not get that there either.  What am I missing ?  Can you please help pointing out ?

**Comment by sjimdickens** [comment]
Hi Venkat,

Thank you for your comments.
1. Yes, the bridging functionality is pointless for now, unless we provide a watch SDK.  But its not a big change for to support, so I have put it in for now.  I delivered a change to fix redact. Thanks for pointing it out. 
2. Fixed the order.

**Comment by sjimdickens** [comment]
Regarding the priority and visibility.  
1. Visibility is set on the application on how the notification is to be displayed when the screen is locked/unlocked. Basically how much information about the notification is to be displayed on the screen. This option is used in combination with redact option.  User can opt to show only part of message on locked screen specified in the redact parameter, and the user needs to click on the notification to see the whole notification. 
2. Priority is used for heads up notification - basically interactive notifications.  When user sets high priority for the notification along with the sound and the user can act on the notification directly.  See below for more information.

https://developer.android.com/guide/topics/ui/notifiers/notifications.html#Heads-up

**Comment by svkrish** [comment]
- 1.  MFPPush.java->sendUpStreamMessage 
  This is UpStreamMessage sent for DISMISS_NOTIFICATION.   Instead of generically calling it sendUpStreamMessage would be good if we clearly call 'relayNotificationDismiss' or 'syncNotificationDismiss'.  Also when you log errors make it specific so that we know what failed is the DISMISS_NOTIFICATION call.   I suppose there can be other kinds of upstream messages that can be sent
- 2) MFPPushIntentService.java
- i) Line 162 : Why is the resource name for notification title '.png' ?
  Why would we not set the notification title as part of the server 'sendMessage' options and then override it with what we might find in the client ?  For WebPush we send the notification title from the server and we could make this uniform for mobile and web.   @srinivasannanduri, what do you think ?
- ii) For the gcmStyle.type constants why are not using enumerations ?   We are using strings which will eat up into the data bandwidth and actually restrict the payload size for users.  
- iii) line 214.  Please be more explicit on what image had issues in loading... in this case it is the image associated with BigPicture notification.  
- iv) so I see we setup whatever the server had sent.  But do we allow the app to override this - I don't see that happening.  i.e. assume we have something like MFPPush.setOptions(BigPictureStyle) and we check if that is set and use it first.. if that is not set .. we use what the server sends.  
- v) all of what I have mentioned above for BigPicuture also applies for BIG_TEXT too.
- vi) INBOX_STYLE.  Same comments as for BigPicture.  In addition am not sure if this makes sense to be sent from the server - it would look like a multi-line notification - sounds very silly to me :).   Does make sense for client side overriding though, whereby a certain category of messages can be grouped and shown in multiple lines.  Maybe we should discuss this a bit.  
  vii) For Priority we have enum on the server which is the perfect thing but I guess we pass the string values of the enums and end up comparing strings on the client.  Why don't we just pass over the intergers and compare integers on the client side too.   Is this something we want the client developer to override ?  @srinivasannanduri ??  I don't see it at the moment and I suppose this will need a bit of design of the hook we should provide to users where they receive notifications and decide to set the priority and then we dispatch it to the NotificationManager.
- viii) For Visibility I have the same comments as I have for Priority.
- ix) Android version specific handling where you are enabling certain features based on the android version - from how it is... it seems like if (message.getGcmStyle() != null && androidSDKVersion > 21) then priority and visibility options will not apply ??  Am I missing anything ?  I wonder if you had a stack of if blocks in descending order would that be better ?
- x)  private Intent handleMessageIntent(Intent intent, Bundle extras) 
  You are handling dismissNotification here.  I think it would be better isolated if you do it in onHandleIntent so that when you get to this method it is, as the name suggests, only about handling the message action.
- xi) protected void dismissNotification(String nid) {
  Ok, this one am not sure of what am saying here - see if it makes sense or just ignore
  I see you are using 'nid' to identify the notification to dismiss.  But this 'nid' is our id right and if I remember this correctly it is a part of the data of a notification.  Will GCM send the data even for dismiss calls  OR would it send the notification key or something for you to correlate.  I think we have to correlate based on what GCM sends and not depend on our metadata like 'nid'.

In summary I see code here to handle... 
- Priority & Heads-up notification which is an extendable behaviour over priority
- Visibility
- Notification Styles (Picture, Text, Inbox)
- Synchronizing notification dismisses
- Loading of custom icons
  If there is anything else, then I totally missed :) - apologies!
  @srinivasannanduri mentioned that he'd like everything customizable for the android app developer i.e. the developer should be able to override these.   I did not observe that.  I only observed that whatever is send from the server is handled appropriately and set - no overrriding.  Am I missing this ?

**Comment by sjimdickens** [comment]
My comments below in this pen </jdsa>

MFPPush.java->sendUpStreamMessage This is UpStreamMessage sent for DISMISS_NOTIFICATION. Instead of generically calling it sendUpStreamMessage would be good if we clearly call 'relayNotificationDismiss' or 'syncNotificationDismiss'. Also when you log errors make it specific so that we know what failed is the DISMISS_NOTIFICATION call. I suppose there can be other kinds of upstream messages that can be sent <jdsa> For now, this is the only upstream call we make. I'll change the names accordingly" <jdsa> 2) MFPPushIntentService.java i) Line 162 : Why is the resource name for notification title '.png' ? Why would we not set the notification title as part of the server 'sendMessage' options and then override it with what we might find in the client ? For WebPush we send the notification title from the server and we could make this uniform for mobile and web. @srinivasannanduri
, what do you think ? **<jdsa> This is a oversight, old code that needs to be cleaned up. I'll remove it </jdsa>** ii) For the gcmStyle.type constants why are not using enumerations ? We are using strings which will eat up into the data bandwidth and actually restrict the payload size for users.
iii) line 214. Please be more explicit on what image had issues in loading... in this case it is the image associated with BigPicture notification.
iv) so I see we setup whatever the server had sent. But do we allow the app to override this - I don't see that happening. i.e. assume we have something like MFPPush.setOptions(BigPictureStyle) and we check if that is set and use it first.. if that is not set .. we use what the server sends. **<jdsa> The expandable notification is an on-the-fly solution to display notifications in different format.  Overriding this option on the client makes no sense because the backend user will selectively send only certain notifications in expandable format. Also these notifications require additional information to be rendered properly on the client, which is sent as part of the message payload. So no overriding here </jdsa>**
v) all of what I have mentioned above for BigPicuture also applies for BIG_TEXT too. vi) INBOX_STYLE. Same comments as for BigPicture. In addition am not sure if this makes sense to be sent from the server - it would look like a multi-line notification - sounds very silly to me :). Does make sense for client side overriding though, whereby a certain category of messages can be grouped and shown in multiple lines. Maybe we should discuss this a bit. **<jdsa> What you are talking about here is "bundling notifications", wherein we group messages of similar category all be grouped under one and display as bundle - involves more than one message.  This feature is something new with Android N, and will be provided for Android N and later.  Keep in mind, expandable notifications are used to provide different visual rendering </jdsa>**
vii) For Priority we have enum on the server which is the perfect thing but I guess we pass the string values of the enums and end up comparing strings on the client. Why don't we just pass over the intergers and compare integers on the client side too. Is this something we want the client developer to override ? @srinivasannanduri ?? I don't see it at the moment and I suppose this will need a bit of design of the hook we should provide to users where they receive notifications and decide to set the priority and then we dispatch it to the NotificationManager. viii) For Visibility I have the same comments as I have for Priority.**<jdsa> We currently send this information from server as string, although we collect as enum.  This will require extensive fixing on the on-prem side as well. We need to factor that too, if we plan on taking this. </jdsa>**  ix) Android version specific handling where you are enabling certain features based on the android version - from how it is... it seems like if (message.getGcmStyle() != null && androidSDKVersion > 21) then priority and visibility options will not apply ?? Am I missing anything ? I wonder if you had a stack of if blocks in descending order would that be better ?**<jdsa>The visibility and priority options won't apply for expandable notifications.  The usage of these notifications are primarily for say ad display via notifications and not to pass any time sensitive or important information along  </jdsa>** x) private Intent handleMessageIntent(Intent intent, Bundle extras) You are handling dismissNotification here. I think it would be better isolated if you do it in onHandleIntent so that when you get to this method it is, as the name suggests, only about handling the message action. xi) protected void dismissNotification(String nid) { Ok, this one am not sure of what am saying here - see if it makes sense or just ignore I see you are using 'nid' to identify the notification to dismiss. But this 'nid' is our id right and if I remember this correctly it is a part of the data of a notification. Will GCM send the data even for dismiss calls OR would it send the notification key or something for you to correlate. I think we have to correlate based on what GCM sends and not depend on our metadata like 'did'. **<jdsa> We store this nid information in our preferences, and we handle clearing notifications on multiple devices. So we are good here. Nothing to do with GCM </jdsa>**
